File: traverse.py
class TNode:
    def __init__(self,data,left,right):
        self.data =data
        self.left =left
        self.right=right

def preorder(p):
    if p is None:
        return
    result.append(p.data)
    preorder(p.left)
    preorder(p.right)


def inorder(p):
    if p is None:
        return
    inorder(p.left)
    result.append(p.data)
    inorder(p.right)
    
def postorder(p):
    if p is None:
        return
    postorder(p.left)
    postorder(p.right)
    result.append(p.data)
    
def solonode(p):
    if p is None:
        return
    solonode(p.left)
    if p.left is None and p.right is None:
        result.append(p.data)
    solonode(p.right)




def printf(inp):
    global result
    global sum
    result =[]
    if inp == 1:
        print("Pre : ",end=' ')
        size = int(input("얼마나 보다 커야하는가?"))
        preorder(a)
        for i in result:
            if i >size:
                print(i,end=' ')



    elif inp == 2:
        print("In : ",end=' ')
        solonode(a)
        for i in result:
            sum += i
        
        print(sum)


        result =[]
    elif inp == 3:
        print("Post : ",end =' ')
        postorder(a)
        print(result)
        print(len(result))

    else:
        print("잘못된 입력입니다.")
        printf(int(input("1. Preorder 2. Inorder 3. Postorder : ")))

sum = 0
result = []

if __name__ == '__main__':
    f = TNode(25, None, None)
    e = TNode(17,None,None)
    d = TNode(4,None,None)
    c = TNode(20,e,f)
    b = TNode(8,d,None)
    a = TNode(15,b,c)

    print("루트 값 :",a.data,"루트의 자식 노드 :",a.left.data,a.right.data)
    print()
    printf(int(input("1. Preorder 2. Inorder 3. Postorder : ")))
    printf(int(input("1. Preorder 2. Inorder 3. Postorder : ")))
    printf(int(input("1. Preorder 2. Inorder 3. Postorder : ")))
    '''print("루트 값 :",a.data,"루트의 자식 노드 :",a.left,a.right)
    
    print()
    print("Pre : ",end=' ')
    preorder(a)
    print("In",end=' ')
    inorder(a)
    print("Post",end =' ')
    postorder(a)'''

File: traverse.c
#include <stdio.h>
#include <stdlib.h>
#include <memory.h>

typedef struct tree_node {
    int data;

    struct tree_node *left;
    struct tree_node *right;
}treenode;
int sum = 0;
int count = 0;

void preorder(treenode *root){
    if (root==NULL){
        return ;
    }
    printf("%d ",root->data);
    preorder(root->left);
    preorder(root->right);
}
void in(treenode *root){
    if (root==NULL){
        return;
    }
    in(root->left);
    if (root->left == NULL && root->right == NULL){
        sum += root -> data;
    }
    in(root->right);
}

void postorder(treenode *root){
    if (root ==NULL){
        return;
    }
    postorder(root->left);
    postorder(root->right);
    count++;


}
void pre(treenode *root){
    if (root==NULL){
        return ;
    }
    if (root->data > 16) {
        printf("%d ", root->data);
    }
    pre(root->left);
    pre(root->right);
}
int main(){
    treenode *A,*B,*C,*D,*E,*F;
    A = (treenode*)malloc(sizeof(treenode));
    B = (treenode*)malloc(sizeof(treenode));
    C = (treenode*)malloc(sizeof(treenode));
    D = (treenode*)malloc(sizeof(treenode));
    E = (treenode*)malloc(sizeof(treenode));
    F = (treenode*)malloc(sizeof(treenode));

    A->data =15;
    A->left =B;
    A->right =C;

    B->data =8;
    B->left =D;
    B->right=NULL;

    C->data =20;
    C->left =E;
    C->right=F;

    D->data =4;
    D->left =NULL;
    D->right =NULL;

    E->data = 17;
    E->left = NULL;
    E->right = NULL;

    F->data=25;
    F->left = NULL;
    F->right = NULL;

printf("%d %d %d\n",A->data,A->left->data,A->right->data);
printf("전위 순회 : ");
pre(A);
printf("\n");
printf("중위 순회 : ");
in(A);
printf("%d ", sum);
printf("\n");
printf("후위 순회 : ");
postorder(A);
printf("%d ", count);
free(A);
free(B);
free(C);
free(D);
free(E);
free(F);
}